homeassistant/components/mqtt/__init__.py

```diff
@@ -683,6 +684,40 @@ def remove():
         DOMAIN, SERVICE_PUBLISH, async_publish_service, schema=MQTT_PUBLISH_SCHEMA
     )
 
+    async def async_dump_service(call: ServiceCall):
+        """Handle MQTT dump service calls."""
+        messages = []
+
+        @callback
+        def collect_msg(msg):
+            messages.append((msg.topic, msg.payload.replace("\n", "")))
+
+        unsub = await async_subscribe(hass, call.data["topic"], collect_msg)
+
+        def write_dump():
+            with open(hass.config.path("mqtt_dump.txt"), "wt") as fp:
+                for msg in messages:
```

Comment:
We should clear the messages somewhere here, right?

Reply:
Once the outer scope gets out of memory the list will be garbage collected.

Reply:
Yes, you're right.